### ISSUE TITLE
Investigate how to approach refactoring of UserSession handling

### ISSUE THREAD

**Issue by mhajas** [keycloak/keycloak#13106]
It is necessary to start a discussion regarding refactoring user session handling. The main reason is that legacy Infinispan storage worked in a different way. Currently, we have a lot of code that can be adjusted/removed because the new store has a different approach. The main points:

- Expiration - new store computes one expiration field/old store has lifespan/max idle. Also new store never returns expired sessions therefore we may be able to remove some validity checks that are part of, for example, `AuthenticationManager`.
- Offline/Online - the old store had methods like `createOnline/Offline` client session, possibly we can have only one method and use the offline flag from a user session. Maybe we can even remove the offline flag from `ClientSessions`.
- Method cleanup - as stated above, a lot of methods/parameters doesn't make sense anymore

Feel free to add more things to consider before starting a discussion

**Comment by sschu** [comment]
In the context of lazy loading offline session @stianst mentioned he thinks it makes more sense to have a 1:1 relation between user and client sessions for offline session since the lifecycle of two offline client sessions created from the same browser session is completely independent (e.g. if you think about revocation).

**Comment by ahus1** [comment]
As announced to the community in [this blog post](https://www.keycloak.org/2023/10/map-store-removal), we are no longer developing the Map Store, and it will be removed from Keycloak's code base. Follow #24111 for the removal steps.

**Comment by ahus1** [comment]
@mhajas - if you can make a case why to change the legacy store to have a positive effect to users or developers, please create a new issue(s). I assume this will make it more focused to understand the benefits than keeping this issue which mostly compares old and new store, but not naming the benefits.

**Comment by mhajas** [comment]
The benefit would be a cleaner code and easier maintenance. I would not say this issue is mostly about comparing new and old stores. The comparing part of this issue is the expiration computation. The other two points can be considered possible refactorings of the user session interface that is relevant regardless of the store implementation.

I cannot say all the details from the top of my head. This would need some more investigation since it is already more than a year since this issue was created but I think it is worth it when we dedicate some time for refactorings.

**Comment by ahus1** [comment]
IMHO we need more information than is currently in the summary to assess this in the team to prioritize it against other issues in our backlog. 

I now re-opened and moved it to "legacy store maintenance", as I assume we don't have the time at our hands now the triage / scope / estimate those refactorings for the legacy store.
